### PR TITLE
add busco tool

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 1]
```diff
@@ -62,10 +67,10 @@
 # if no specific metadata is provided for a tool, the default values are used.
 TASK_META = {'gtdb_tk': {'chunk_size': 1000, 'exe_time': 65, 'tasks_per_node': 4, 'threads_per_tool_run': 32},
              'eggnog': {'chunk_size': 100, 'exe_time': 15, 'node_time_limit': 0.5},  # Memory intensive tool - reserve more nodes with less node reservation time
+             'busco': {'chunk_size': 50, 'exe_time': 90, 'node_time_limit': 1.5},  # 1.5 minutes per genome with a single task per node on the user's drive. TODO: Aim to test multi-threading per node along with scratch execution, and adjust `tasks_per_node` accordingly.
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:70]: Wow this is expensive. For 10k genomes that's 250 hours, so scaling this to GTDB sized genome counts seems infeasible. Hopefully your TODOs pan out.

File: src/loaders/compute_tools/busco/versions.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# This tool serves a distinct purpose separate from collection tools; instead, it is suited for CDM work.
+# Therefore, the parser program is not compatible with data generated by this tool.
+
+versions:
+  - version: 0.1.0
+    date: 2024-08-22
+    notes: |
+      - initial BUSCO implementation
+    reference_db_version: odb10
```
Comment (MrCreosote) [inline src/loaders/compute_tools/busco/versions.yaml]: Remind me, ref data is installed manually right?

File: src/loaders/compute_tools/busco/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# Add Bioconda and Conda-Forge channels
+RUN conda config --add channels bioconda
+RUN conda config --add channels conda-forge
+
+# Install BUSCO
+# Certain dependencies (e.g., dendropy, sepp) are only compatible with Python versions up to 3.9.
+ARG PYTHON_VER=3.9
+RUN conda create -n $CONDA_ENV python=$PYTHON_VER
+RUN conda install -n $CONDA_ENV pandas=2.2.2 jsonlines=4.0.0 mamba=1.5.8 pyyaml=6.0.1
+RUN conda run -n $CONDA_ENV mamba install -c bioconda -c conda-forge -y busco=$BUSCO_VER
```
Comment (MrCreosote) [inline src/loaders/compute_tools/busco/Dockerfile:18]: I'm not familiar enough with conda / mamba to understand exactly what's going on here, can you explain?

File: src/loaders/compute_tools/busco/busco.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    run_time = end_time - start
+    print(
+        f'Used {round(run_time / 60, 2)} minutes to execute BUSCO for {data_id}')
+
+    # Save run info to a metadata file in the output directory for parsing later
+    additional_metadata = {
+        'source_file': str(source_file),
+        'data_id': data_id,
+        "reference_db": {
+            "version": "odb10",
```
Comment (MrCreosote) [inline src/loaders/compute_tools/busco/busco.py]:
```suggestion
            "version": ref_db_version,
```

Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py]: I think this means you could also delete the note about updating the ref db in the version yaml

File: src/loaders/compute_tools/busco/busco.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    ref_db_version = extract_latest_reference_db_version(version_file)
+
+    command = [
+        'busco',
+        '-i', str(source_file),
+        '-o', data_id,
+        '--out_path', str(output_dir),
+        '--datasets_version', ref_db_version,
+        '--download_path', '/reference_data',
+        '-c', str(threads_per_tool_run),
```
Comment (MrCreosote) [inline src/loaders/compute_tools/busco/busco.py:45]: I forget how this gets set, can you remind me? Looking at the task generator changes it's not clear.

File: src/loaders/compute_tools/busco/busco.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        '--auto-lineage-prok',
+        '-m', 'genome',
+        '-f',
+        '--augustus',
```
Comment (MrCreosote) [inline src/loaders/compute_tools/busco/busco.py:49]: brief inline comments as to what these args do would be helpful

File: src/loaders/compute_tools/busco/busco.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/busco.py]: 👍

File: src/loaders/compute_tools/busco/Dockerfile (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/Dockerfile:18]: added comments.

File: src/loaders/compute_tools/busco/busco.py (hunk 5 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/busco.py:45]: https://github.com/kbase/collections/blob/develop/src/loaders/jobs/taskfarmer/task_generator.py#L59

File: src/loaders/compute_tools/busco/busco.py (hunk 6 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/busco.py:49]: added comments.

File: src/loaders/compute_tools/busco/versions.yaml (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/versions.yaml]: yea.

Comment (MrCreosote) [reply src/loaders/compute_tools/busco/versions.yaml]: :+1:

File: src/loaders/compute_tools/busco/busco.py (hunk 5 above)
Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py:45]: Ok, so BUSCO is using 32 threads to process a single genome in 90sec? Do different thread counts affect the processing time?

File: src/loaders/compute_tools/busco/Dockerfile (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/compute_tools/busco/Dockerfile:18]: Why the `conda run`? The linked instructions have mamba executed separately

File: src/loaders/compute_tools/busco/busco.py (hunk 6 above)
Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py:49]: :+1:

File: src/loaders/compute_tools/busco/busco.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py]: :+1:

File: src/loaders/compute_tools/busco/Dockerfile (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/Dockerfile:18]: to ensure mamba is called from the correct conda environment.

Comment (MrCreosote) [reply src/loaders/compute_tools/busco/Dockerfile:18]: :+1:

File: src/loaders/compute_tools/busco/busco.py (hunk 5 above)
Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/busco.py:45]: updated the outdated comment. It's using 256 threads by default.

Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py:45]: Ok, so one of the TODOs is seeing if you can run multiple instances with fewer threads each per node to reduce the allocation used?

Comment (Tianhao-Gu) [reply src/loaders/compute_tools/busco/busco.py:45]: yea. If we were running more tasks per node, the threads per task will be automatically determined using the `SYSTEM_CPU_CORES / number of tasks` by default.

Comment (MrCreosote) [reply src/loaders/compute_tools/busco/busco.py:45]: :+1:

**Review by MrCreosote — APPROVED** [review]
LGTM

**Pull request merged**